File: nodejs/05-linearSorting/linearSort.js
/**
 * Counting Sort Algorithm
 * @param {number[]} arr - The array to be sorted
 * @returns {number[]} - The sorted array
 */
function countingSort(arr) {
  if (arr.length === 0) return [];

  // Find the maximum and minimum values in the array
  const max = Math.max(...arr);
  const min = Math.min(...arr);

  // Create a count array to store the count of each unique number
  const countArray = new Array(max - min + 1).fill(0);

  // Count the occurrences of each number
  for (let num of arr) {
      countArray[num - min]++;
  }

  // Build the sorted array
  let sortedIndex = 0;
  for (let i = 0; i < countArray.length; i++) {
      while (countArray[i] > 0) {
          arr[sortedIndex++] = i + min;
          countArray[i]--;
      }
  }

  return arr;
}



/**
 * Helper function to get the digit at a specific place value
 * @param {number} num - The number from which to extract the digit
 * @param {number} place - The place value (0 for units, 1 for tens, etc.)
 * @returns {number} - The digit at the specified place value
 */
function getDigit(num, place) {
  return Math.floor(Math.abs(num) / Math.pow(10, place)) % 10;
}

/**
* Helper function to count the number of digits in a number
* @param {number} num - The number to count digits
* @returns {number} - The number of digits in the number
*/
function digitCount(num) {
  if (num === 0) return 1;
  return Math.floor(Math.log10(Math.abs(num))) + 1;
}

/**
* Helper function to find the maximum number of digits in an array of numbers
* @param {number[]} nums - The array of numbers
* @returns {number} - The maximum number of digits found in the array
*/
function mostDigits(nums) {
  let maxDigits = 0;
  for (let num of nums) {
      maxDigits = Math.max(maxDigits, digitCount(num));
  }
  return maxDigits;
}

/**
* Radix Sort Algorithm
* @param {number[]} nums - The array of numbers to be sorted
* @returns {number[]} - The sorted array
*/
function radixSort(nums) {
  const maxDigitCount = mostDigits(nums);

  // Loop through each digit place (units, tens, hundreds, etc.)
  for (let k = 0; k < maxDigitCount; k++) {
      // Create buckets for each digit (0 to 9)
      const digitBuckets = Array.from({ length: 10 }, () => []);

      // Place each number in the corresponding bucket based on the current digit
      for (let num of nums) {
          const digit = getDigit(num, k);
          digitBuckets[digit].push(num);
      }

      // Flatten the buckets back into the array
      nums = [].concat(...digitBuckets);
  }

  return nums;
}

File: linearSorting_05/linearSort.py
from typing import List

def counting_sort(arr: List[int]) -> List[int]:
    """
    Counting Sort Algorithm

    :param arr: List of integers to be sorted
    :return: Sorted list of integers
    """
    if not arr:
        return []

    # Find the maximum and minimum values in the array
    max_val = max(arr)
    min_val = min(arr)

    # Create a count array to store the count of each unique number
    count_array = [0] * (max_val - min_val + 1)

    # Count the occurrences of each number
    for num in arr:
        count_array[num - min_val] += 1

    # Build the sorted array
    sorted_index = 0
    for i in range(len(count_array)):
        while count_array[i] > 0:
            arr[sorted_index] = i + min_val
            sorted_index += 1
            count_array[i] -= 1

    return arr

from typing import List

def get_digit(num: int, place: int) -> int:
    """
    Get the digit at a specific place value

    :param num: The number from which to extract the digit
    :param place: The place value (0 for units, 1 for tens, etc.)
    :return: The digit at the specified place value
    """
    return abs(num) // (10 ** place) % 10

def digit_count(num: int) -> int:
    """
    Count the number of digits in a number

    :param num: The number to count digits
    :return: The number of digits in the number
    """
    if num == 0:
        return 1
    return len(str(abs(num)))

def most_digits(nums: List[int]) -> int:
    """
    Find the maximum number of digits in an array of numbers

    :param nums: The array of numbers
    :return: The maximum number of digits found in the array
    """
    max_digits = 0
    for num in nums:
        max_digits = max(max_digits, digit_count(num))
    return max_digits

def radix_sort(nums: List[int]) -> List[int]:
    """
    Radix Sort Algorithm

    :param nums: List of integers to be sorted
    :return: Sorted list of integers
    """
    max_digit_count = most_digits(nums)

    # Loop through each digit place (units, tens, hundreds, etc.)
    for k in range(max_digit_count):
        # Create buckets for each digit (0 to 9)
        digit_buckets = [[] for _ in range(10)]

        # Place each number in the corresponding bucket based on the current digit
        for num in nums:
            digit = get_digit(num, k)
            digit_buckets[digit].append(num)

        # Flatten the buckets back into the array
        nums = [num for bucket in digit_buckets for num in bucket]

    return nums

# Example usage
arr2 = [170, 45, 75, 90, 802, 24, 2, 66]
print(radix_sort(arr2))  # Output: [2, 24, 45, 66, 75, 90, 170, 802]


# Example usage
arr1 = [4, 2, 2, 8, 3, 3, 1]
print(counting_sort(arr1))  # Output: [1, 2, 2, 3, 3, 4, 8]
